### PR TITLE
Changes for Issue 91

### PR DESCRIPTION
**Pull request by davetrig**
Changes for Issue 91 - Automatically cache export after creating new repository version

Added _ocl_processing field to CollectionVersion class.  This field was in SourceVersion, but not in CollectionVersion.  Oddly, I did see places in API code that referred to the CollectionVersion._ocl_processing field, but I could not find an explicit declaration, so I added it just to be sure.

On GET and POST for source and collection version export, if the _ocl_processing field of the requested version is True, HTTP status 409 (Conflict) is returned.

Upon source or collection version creation, the appropriate task is called to create the export.

@rkorytkowski and @paynejd:
I also created integration tests to verify the return of status 409 when version._ocl_processing is true.  However, I discovered what I believe to be a major flaw in the existing integration tests.  Almost all of the POST tests for export creation are actually doing GETs.  The only tests doing POSTs are those testing HEAD version, or testing version ids that do not exist.  When I attempt a POST on a version that does exist, the POST returns "401: Authentication credentials were not provided", no matter what method I try to provide credentials.  

I also see that in the existing POST tests, authentication is ostensibly done with a POST to "/login/"; however, every one of these posts returns a 404 error, as there is no matching URL pattern.  I attempted my authentication with "/auth-api/login/", which returns status 200, but still get a 401 error when POSTing the version export request.

I have, for the meanwhile, followed the existing pattern of issuing a GET rather than a POST, but this is an issue we should figure out at some point, because these integration tests are not actually testing their respective functionality.

### CODE REVIEW

**Comment by rkorytkowski** [comment]
Dave, did you try posting with the http "Authorization" header set to "Token 891b4b17feab99f3ff7e5b5d04ccc5da7aa96da6"?

File: django-nonrel/ocl/integration_tests/tests/all.py [hunk 1]
```diff
@@ -2108,6 +2108,95 @@ def test_post(self):
...
+            versioned_object=source,
+            released=True,
+            created_by=self.user1,
+            updated_by=self.user1,
+            _ocl_processing=True
+        )
+        source_version.full_clean()
+        source_version.save()
+        kwargs = {'parent_resource': source}
+        (concept1, _) = create_concept(mnemonic='concept1', user=self.user1, source=source)
```
Comment (rkorytkowski) [inline django-nonrel/ocl/integration_tests/tests/all.py]: Why is concept being created here?

File: django-nonrel/ocl/integration_tests/tests/all.py [hunk 2]
```diff
@@ -2108,6 +2108,95 @@ def test_post(self):
...
+        c = Client()
+        c.post('/login/', {'username': 'user1', 'password': 'user1'})
+
+        kwargs = {
+            'org': self.org1.mnemonic,
+            'source': source.mnemonic,
+            'version': 'version1'
+        }
+        response = c.get(reverse('sourceversion-export', kwargs=kwargs))
+        self.assertEquals(response.status_code, 409)
```
Comment (rkorytkowski) [inline django-nonrel/ocl/integration_tests/tests/all.py]: Is this going to be a stable test? Is it possible that the export is ready by the time you do get and it returns 204?

**Comment by rkorytkowski** [comment]
The code looks good. I left some questions on tests. I'll merge it in to do manual testing.

File: django-nonrel/ocl/integration_tests/tests/all.py (hunk 2 above)
Comment (davetrig) [reply django-nonrel/ocl/integration_tests/tests/all.py]: There's no export, it's creating the version with _ocl_processing = True, as though the version is still being created.  Therefore, it should always respond with a 409.

File: django-nonrel/ocl/integration_tests/tests/all.py (hunk 1 above)
Comment (davetrig) [reply django-nonrel/ocl/integration_tests/tests/all.py]: No reason - just copy/paste from another test.

**Comment by davetrig** [comment]
> Dave, did you try posting with the http "Authorization" header set to "Token 891b4b17feab99f3ff7e5b5d04ccc5da7aa96da6"?

@rkorytkowski, where are you getting that auth key from?  I don't see it used in any other tests, nor any examples of tests authenticating in this way.

File: django-nonrel/ocl/integration_tests/tests/all.py (hunk 2 above)
Comment (rkorytkowski) [reply django-nonrel/ocl/integration_tests/tests/all.py]: Okay!

**Comment by rkorytkowski** [comment]
It's the api key... REST clients are supposed to authenticate that way, not sure why tests are not using this method. You should be able to get the proper value from the test user object.

**Comment by davetrig** [comment]
As I said, *none* of the other POST tests are authenticating properly, so it doesn't surprise me that no other tests are doing it that way. 
The integration test process creates its own test db, then destroys it, right? So wouldn't the api key change every time?  Or am I thinking of this in the wrong way?

**Comment by davetrig** [comment]
@rkorytkowski, assuming your intention is that I retrieve the proper key during the test, can you give me a code example that retrieves the proper key, and adds it to the header? Then I'll see about adding it to the test in an elegant and reusable way.

**Comment by rkorytkowski** [comment]
Yes, it's exactly what I meant. I'll try to code it for you, but first could you please try to call client.login e.g. https://github.com/OpenConceptLab/oclapi/blob/master/django-nonrel/ocl/integration_tests/tests/openmrs_concept_validation.py#L61-L71 ?

**Comment by davetrig** [comment]
I did actually try it with client.login, which also did not work.  I can try it again to be sure.
You don't need to code the whole solution; I could just use a point in the right direction to the correct functions to call.  It would save me the time trying to figure it out.

**Comment by davetrig** [comment]
I again tried using client.login to authenticate, just to make sure I tried it right the first time.  Still get 401 response from the subsequent POST.

**Comment by davetrig** [comment]
@rkorytkowski, I may have spoken too soon. I tried client.login, but I just realized that nothing in the tests is actually creating the user.  I'm going to try it again, and let you know.

**Comment by davetrig** [comment]
@rkorytkowski I solved it.
There's nothing wrong with the authentication method.  client.login works fine.  The problem was with the way the version was saved.

The collection version was being saved in this way:
`collection_version.full_clean()`
`collection_version.save()`

I changed it to:
`CollectionVersion.persist_new(collection_version, self.user1)`
...and all is right with the world.  I'll make the changes and submit a pull request.

**Pull request merged**